From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 1 Jun 2022 14:01:39 +1000
Subject: [PATCH] Add support for proxy protocol


diff --git a/src/main/java/puregero/multipaper/ExternalServerConnection.java b/src/main/java/puregero/multipaper/ExternalServerConnection.java
index 7fc7f967ff3cad2615ee2f45c5d29ea992296f07..f70a5cb955ef9682a60cd7816ed93e381ca6a59e 100644
--- a/src/main/java/puregero/multipaper/ExternalServerConnection.java
+++ b/src/main/java/puregero/multipaper/ExternalServerConnection.java
@@ -1,27 +1,30 @@
 package puregero.multipaper;
 
 import io.netty.channel.Channel;
+import io.netty.channel.ChannelFuture;
 import io.netty.channel.ChannelFutureListener;
 import io.netty.channel.ChannelInitializer;
 import io.netty.channel.socket.SocketChannel;
 import io.netty.handler.codec.compression.Zstd;
+import io.netty.handler.codec.haproxy.*;
 import io.netty.util.internal.SystemPropertyUtil;
 import net.minecraft.network.PacketEncoder;
 import net.minecraft.network.protocol.Packet;
 import net.minecraft.server.MinecraftServer;
 import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.network.protocol.handshake.ClientIntent;
 import net.minecraft.network.protocol.handshake.ClientIntentionPacket;
 import net.minecraft.network.protocol.handshake.HandshakeProtocols;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 import puregero.multipaper.config.MultiPaperConfiguration;
 import puregero.multipaper.externalserverprotocol.*;
 import puregero.multipaper.mastermessagingprotocol.MessageBootstrap;
 import puregero.multipaper.mastermessagingprotocol.MessageLengthDecoder;
 import puregero.multipaper.mastermessagingprotocol.MessageLengthEncoder;
 
 import java.io.Closeable;
 import java.io.IOException;
 import java.util.*;
+import java.util.concurrent.TimeUnit;
 import java.util.concurrent.CompletableFuture;
@@ -65,15 +65,34 @@ public class ExternalServerConnection extends ChannelInitializer<SocketChannel>
     public void sendMinecraftHandshake(String address, String secret, int port) {
+        // MultiPaper start - Add support for Proxy Protocol
+        if (io.papermc.paper.configuration.GlobalConfiguration.get() == null) {
+            // Paper config hasn't been loaded yet, try again 100ms later
+            CompletableFuture.delayedExecutor(100, TimeUnit.MILLISECONDS).execute(() -> sendMinecraftHandshake(address, secret, port));
+            return;
+        }
+
+        ChannelFuture future2 = channel.newSucceededFuture();
+
+        if (io.papermc.paper.configuration.GlobalConfiguration.get().proxies.proxyProtocol) {
+            channel.pipeline().addLast("haproxy-encoder", HAProxyMessageEncoder.INSTANCE);
+            future2 = channel.writeAndFlush(new HAProxyMessage(HAProxyProtocolVersion.V2, HAProxyCommand.LOCAL, HAProxyProxiedProtocol.UNKNOWN, null, null, 0, 0));
+        }
+
+        future2.addListener(future1 -> {
+        // MultiPaper end
+
         channel.pipeline()
                 .addLast("prepender", new MessageLengthEncoder())
                 .addLast("encoder", new PacketEncoder(HandshakeProtocols.SERVERBOUND));
         channel.writeAndFlush(new ClientIntentionPacket(0, address + "\00" + secret, port, ClientIntent.STATUS))
                 .addListener(future -> {
                     if (future.isSuccess()) {
                         setupPipeline();
                     } else if (future.cause() != null) {
                         future.cause().printStackTrace();
                     }
                 });
+
+        }); // MultiPaper - Add support for Proxy Protocol
     }
 
     public void setupPipeline() {
