From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Fri, 26 Nov 2021 01:30:05 +1000
Subject: [PATCH] Interact with external entities


diff --git a/src/main/java/puregero/multipaper/MultiPaperEntityInteractHandler.java b/src/main/java/puregero/multipaper/MultiPaperEntityInteractHandler.java
new file mode 100644
index 0000000000000000000000000000000000000000..6b212628a7c4e4bc9d3458fc7a79d88bbf866c47
--- /dev/null
+++ b/src/main/java/puregero/multipaper/MultiPaperEntityInteractHandler.java
@@ -0,0 +1,54 @@
+package puregero.multipaper;
+
+import io.papermc.paper.chunk.system.scheduling.NewChunkHolder;
+import net.minecraft.network.protocol.game.ServerboundInteractPacket;
+import net.minecraft.server.level.ServerPlayer;
+import net.minecraft.world.entity.Entity;
+import net.minecraft.world.entity.player.Player;
+import net.minecraft.world.entity.projectile.FishingHook;
+import net.minecraft.world.level.chunk.LevelChunk;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import puregero.multipaper.externalserverprotocol.PlayerActionOnEntityPacket;
+import puregero.multipaper.externalserverprotocol.PlayerTouchEntityPacket;
+
+public class MultiPaperEntityInteractHandler {
+
+    private static final Logger LOGGER = LogManager.getLogger(MultiPaperEntityInteractHandler.class.getSimpleName());
+
+    public static ExternalServerConnection getOwner(Entity entity) {
+        NewChunkHolder newChunkHolder = MultiPaper.getChunkHolder(entity);
+
+        if (MultiPaperEntitiesHandler.getControllingPassenger(entity) instanceof ExternalPlayer externalPlayer) {
+            return externalPlayer.externalServerConnection;
+        } else if (MultiPaper.isChunkExternal(newChunkHolder) && !MultiPaper.isRealPlayer(MultiPaperEntitiesHandler.getControllingPassenger(entity))) {
+            return newChunkHolder.externalOwner.getConnection();
+        }
+
+        return null;
+    }
+
+    public static boolean handleEntityInteract(ServerPlayer player, Entity entity, ServerboundInteractPacket packet) {
+        ExternalServerConnection owner = getOwner(entity);
+        if (owner != null) {
+            owner.send(new PlayerActionOnEntityPacket(player, entity, packet));
+            return true;
+        }
+
+        return false;
+    }
+
+    public static boolean touchEntity(Player player, Entity entity) {
+        NewChunkHolder newChunkHolder = MultiPaper.getChunkHolder(entity);
+
+        if (MultiPaper.isRealPlayer(entity)) {
+            return false;
+        } else if (MultiPaper.isChunkExternal(newChunkHolder) && !(entity instanceof FishingHook)) {
+            newChunkHolder.externalOwner.getConnection().send(new PlayerTouchEntityPacket(player, entity));
+
+            return true;
+        }
+
+        return false;
+    }
+}
diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java b/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
index 8ce98e04ed1400146b69242a3a28c9689fd47e68..ff5b08a3e27643b5e3153f3f058fd0a86090b347 100644
--- a/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/ExternalServerPacketSerializer.java
@@ -42,6 +42,8 @@ public class ExternalServerPacketSerializer {
         addPacket(EntityUpdateWithDependenciesPacket.class, EntityUpdateWithDependenciesPacket::new);
         addPacket(RequestEntityPacket.class, RequestEntityPacket::new);
         addPacket(EntityRemovePacket.class, EntityRemovePacket::new);
+        addPacket(PlayerActionOnEntityPacket.class, PlayerActionOnEntityPacket::new);
+        addPacket(PlayerTouchEntityPacket.class, PlayerTouchEntityPacket::new);
     }
 
     private static void addPacket(Class<? extends ExternalServerPacket> clazz, Function<FriendlyByteBuf, ExternalServerPacket> deserializer) {
diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/PlayerActionOnEntityPacket.java b/src/main/java/puregero/multipaper/externalserverprotocol/PlayerActionOnEntityPacket.java
new file mode 100644
index 0000000000000000000000000000000000000000..e974cb3cbfca212e3c972d98e4fc6ad48cfc2ded
--- /dev/null
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/PlayerActionOnEntityPacket.java
@@ -0,0 +1,90 @@
+package puregero.multipaper.externalserverprotocol;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+import net.minecraft.network.ConnectionProtocol;
+import net.minecraft.network.FriendlyByteBuf;
+import net.minecraft.network.protocol.Packet;
+import net.minecraft.network.protocol.PacketFlow;
+import net.minecraft.network.protocol.game.ServerGamePacketListener;
+import net.minecraft.network.protocol.game.ServerboundInteractPacket;
+import net.minecraft.server.MinecraftServer;
+import net.minecraft.server.level.ServerLevel;
+import net.minecraft.server.level.ServerPlayer;
+import net.minecraft.world.entity.Entity;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import puregero.multipaper.ExternalServerConnection;
+import puregero.multipaper.MultiPaper;
+
+import java.util.UUID;
+
+public class PlayerActionOnEntityPacket extends ExternalServerPacket {
+
+    private static final Logger LOGGER = LogManager.getLogger(PlayerActionOnEntityPacket.class.getSimpleName());
+
+    private final UUID uuid;
+    private final UUID entityUuid;
+    private final Packet<ServerGamePacketListener> action;
+
+    public PlayerActionOnEntityPacket(ServerPlayer player, Entity entity, Packet<ServerGamePacketListener> action) {
+        this.uuid = player.getUUID();
+        this.entityUuid = entity.getUUID();
+        this.action = action;
+    }
+
+    public PlayerActionOnEntityPacket(FriendlyByteBuf in) {
+        this.uuid = in.readUUID();
+        this.entityUuid = in.readUUID();
+
+        ByteBuf buf = Unpooled.wrappedBuffer(in.readByteArray());
+        FriendlyByteBuf friendlyByteBuf = new FriendlyByteBuf(buf);
+        int packetId = friendlyByteBuf.readVarInt();
+        action = (Packet<ServerGamePacketListener>) ConnectionProtocol.PLAY.createPacket(PacketFlow.SERVERBOUND, packetId, friendlyByteBuf);
+    }
+
+    @Override
+    public void write(FriendlyByteBuf out) {
+        out.writeUUID(uuid);
+        out.writeUUID(entityUuid);
+
+        ConnectionProtocol protocol = ConnectionProtocol.getProtocolForPacket(action);
+        Integer id = protocol.getPacketId(PacketFlow.SERVERBOUND, action);
+        ByteBuf buf = Unpooled.buffer();
+        FriendlyByteBuf friendlyByteBuf = new FriendlyByteBuf(buf);
+        friendlyByteBuf.writeVarInt(id);
+        action.write(friendlyByteBuf);
+        out.writeByteArray(buf.array());
+    }
+
+    @Override
+    public void handle(ExternalServerConnection connection) {
+        MultiPaper.runSync(() -> {
+            ServerPlayer player = MinecraftServer.getServer().getPlayerList().getPlayer(uuid);
+
+            if (player == null) {
+                LOGGER.warn("Tried to run an action on a non-existent player with uuid " + uuid);
+                return;
+            }
+
+            Entity entity = ((ServerLevel) player.level).getEntity(entityUuid);
+
+            if (entity == null) {
+                LOGGER.warn(player.getScoreboardName() + " tried to run an action on a non-existent entity with uuid " + entityUuid);
+                return;
+            }
+
+            Packet<ServerGamePacketListener> newPacket;
+
+            // Refactor the entity id
+            if (action instanceof ServerboundInteractPacket serverboundInteractPacket) {
+                newPacket = new ServerboundInteractPacket(entity.getId(), serverboundInteractPacket.isUsingSecondaryAction(), serverboundInteractPacket.getAction());
+            } else {
+                LOGGER.error("Unhandled action on entity " + action);
+                return;
+            }
+
+            newPacket.handle(player.connection);
+        });
+    }
+}
diff --git a/src/main/java/puregero/multipaper/externalserverprotocol/PlayerTouchEntityPacket.java b/src/main/java/puregero/multipaper/externalserverprotocol/PlayerTouchEntityPacket.java
new file mode 100644
index 0000000000000000000000000000000000000000..445683cde026f6d6ef5500f2c5139325405d9797
--- /dev/null
+++ b/src/main/java/puregero/multipaper/externalserverprotocol/PlayerTouchEntityPacket.java
@@ -0,0 +1,66 @@
+package puregero.multipaper.externalserverprotocol;
+
+import net.minecraft.network.FriendlyByteBuf;
+import net.minecraft.server.MinecraftServer;
+import net.minecraft.server.level.ServerLevel;
+import net.minecraft.server.level.ServerPlayer;
+import net.minecraft.world.entity.Entity;
+import net.minecraft.world.entity.player.Player;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import puregero.multipaper.ExternalServerConnection;
+import puregero.multipaper.MultiPaper;
+
+import java.util.UUID;
+
+public class PlayerTouchEntityPacket extends ExternalServerPacket {
+
+    private static final Logger LOGGER = LogManager.getLogger(PlayerTouchEntityPacket.class.getSimpleName());
+
+    private final UUID uuid;
+    private final UUID entityUuid;
+
+    public PlayerTouchEntityPacket(Player player, Entity entity) {
+        this.uuid = player.getUUID();
+        this.entityUuid = entity.getUUID();
+    }
+
+    public PlayerTouchEntityPacket(FriendlyByteBuf in) {
+        uuid = in.readUUID();
+        entityUuid = in.readUUID();
+    }
+
+    @Override
+    public void write(FriendlyByteBuf out) {
+        out.writeUUID(uuid);
+        out.writeUUID(entityUuid);
+    }
+
+    @Override
+    public void handle(ExternalServerConnection connection) {
+        MultiPaper.runSync(() -> {
+            ServerPlayer player = MinecraftServer.getServer().getPlayerList().getPlayer(uuid);
+
+            if (player == null) {
+                LOGGER.warn("Tried to run a touch entity on a non-existent player with uuid " + uuid);
+                return;
+            }
+
+            Entity entity = ((ServerLevel) player.level()).getEntity(entityUuid);
+
+            if (entity == null) {
+                Entity.RemovalReason removalReason = EntityRemovePacket.removedEntities.get(entityUuid);
+                if (removalReason != null && removalReason.shouldDestroy()) {
+                    connection.send(new EntityRemovePacket(player.level().getWorld().getName(), entityUuid, false));
+                    return;
+                }
+
+                LOGGER.warn(player.getScoreboardName() + " tried to touch a non-existent entity with uuid " + entityUuid + ", requesting it...");
+                RequestEntityPacket.requestEntity(connection, ((ServerLevel) player.level()).uuid, entityUuid);
+                return;
+            }
+
+            entity.playerTouch(player);
+        });
+    }
+}
