From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Mon, 15 Nov 2021 13:02:02 +1000
Subject: [PATCH] Add sendTickTime


diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index d9102845c6631d75e7549f269ca6b55bf9e8c051..8c536848a4b5910096a6ef78a7c3bfc9336abc2f 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -1468,6 +1468,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         tickTimes10s.add(this.tickCount, j);
         tickTimes60s.add(this.tickCount, j);
         // Paper end
+        MultiPaper.sendTickTime(k - i, recentTps[0]); // MultiPaper
         this.frameTimer.logFrameDuration(k - i);
         //this.profiler.pop(); // Purpur
         org.spigotmc.WatchdogThread.tick(); // Spigot
