From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sat, 4 Dec 2021 12:23:17 +1000
Subject: [PATCH] Sync enderchests


diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index d38fa61c8264442a0712227d20281fbe9396a248..be331455a878093fb12a704dd7d5870706ae9785 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -433,6 +433,7 @@ public class ServerPlayer extends Player {
         this.maxHealthCache = this.getMaxHealth();
         this.cachedSingleMobDistanceMap = new com.destroystokyo.paper.util.PooledHashSets.PooledObjectLinkedOpenHashSet<>(this); // Paper
         this.spawnInvulnerableTime = world.purpurConfig.playerSpawnInvulnerableTicks; // Purpur
+        enderChestInventory.addListener(new MultiPaperEnderChestHandler(this)); // MultiPaper
     }
 
     // Yes, this doesn't match Vanilla, but it's the best we can do for now.
