From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sat, 4 May 2024 15:33:31 +0900
Subject: [PATCH] Add watchdog-like to the shutdown process


diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index d9402a730568d66110a5f87c6a97f8b510c9a3a6..11dc053c192df96b16fb90a2e7b553c81a0bb357 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -1024,6 +1024,42 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
             this.cancelRecordingMetrics();
         }
 
+        // MultiPaper start - add a watchdog to the server stopping
+        Thread shutdownWatchdog = new Thread(() -> {
+            long start = System.currentTimeMillis();
+            int count = 0;
+            while (true) {
+                try {
+                    Thread.sleep(10000);
+                } catch (InterruptedException e) {
+                    e.printStackTrace();
+                }
+                LOGGER.error("The server has not stopped after {} seconds! Creating thread dump", (System.currentTimeMillis() - start) / 1000);
+                if (shutdownThread.isAlive()) {
+                    LOGGER.error("{} Stack:", shutdownThread.getName());
+                    for (StackTraceElement element : shutdownThread.getStackTrace()) {
+                        LOGGER.error("\t{}", element);
+                    }
+                } else {
+                    LOGGER.error("\tThe shutdown thread isn't alive!!!");
+                }
+
+                if ((++count) % 6 == 0) {
+                    LOGGER.error("All thread stacks:");
+                    for (Map.Entry<Thread, StackTraceElement[]> entry : Thread.getAllStackTraces().entrySet()) {
+                        Thread thread = entry.getKey();
+                        LOGGER.error("\t{} Stack:", thread.getName());
+                        for (StackTraceElement element : entry.getValue()) {
+                            LOGGER.error("\t\t{}", element);
+                        }
+                    }
+                }
+            }
+        }, "Stopping-Watchdog");
+        shutdownWatchdog.setDaemon(true);
+        shutdownWatchdog.start();
+        // MultiPaper end
+
         LOGGER.info("Stopping server");
         Commands.COMMAND_SENDING_POOL.shutdownNow(); // Paper - Perf: Async command map building; Shutdown and don't bother finishing
         // Purpur start - UPnP Port Forwarding
