From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Tue, 14 Dec 2021 00:54:24 +1000
Subject: [PATCH] Sync advancements in real time


diff --git a/net/minecraft/server/PlayerAdvancements.java b/net/minecraft/server/PlayerAdvancements.java
index 7c22cd3bc416c04ee7818f870486c9e51b93580c..12f35ba396f9bcc8776f889b239d11a16fdb6842 100644
--- a/net/minecraft/server/PlayerAdvancements.java
+++ b/net/minecraft/server/PlayerAdvancements.java
@@ -47,6 +47,7 @@ import net.minecraft.world.level.GameRules;
 import org.slf4j.Logger;
 import puregero.multipaper.ExternalPlayer;
 import puregero.multipaper.MultiPaper;
+import puregero.multipaper.externalserverprotocol.AdvancementGrantProgressPacket;
 
 public class PlayerAdvancements {
 
@@ -259,10 +260,11 @@ public class PlayerAdvancements {
                 return false;
             }
             // Paper end
+            AdvancementGrantProgressPacket.onAdvancementGrantProgress(this.player, advancement, criterionName);
             this.unregisterListeners(advancement);
             this.progressChanged.add(advancement);
             flag = true;
-            if (!flag1 && advancementprogress.isDone()) {
+            if (!flag1 && advancementprogress.isDone() && MultiPaper.isRealPlayer(player)) { // MultiPaper - only broadcast advancement message for our players
                 // Paper start - Add Adventure message to PlayerAdvancementDoneEvent
                 boolean announceToChat = advancement.getDisplay() != null && advancement.getDisplay().shouldAnnounceChat();
                 net.kyori.adventure.text.Component message = announceToChat ? io.papermc.paper.adventure.PaperAdventure.asAdventure(Component.translatable("chat.type.advancement." + advancement.getDisplay().getFrame().getName(), this.player.getDisplayName(), advancement.getChatComponent())) : null;
