From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Mon, 29 Nov 2021 22:20:38 +1000
Subject: [PATCH] Airplane Configuration


diff --git a/build.gradle.kts b/build.gradle.kts
index 862ad694b2a6b1ed2c7113a9ba0efe1ad62faf1c..df3880a2e7c4250eff1d8567a73aa8e7e0b40f87 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -46,12 +46,14 @@ dependencies {
     runtimeOnly("org.apache.maven.resolver:maven-resolver-transport-http:1.7.3")
 
     // Pufferfish start
-    implementation("org.yaml:snakeyaml:1.33")
-    implementation ("me.carleslc.Simple-YAML:Simple-Yaml:1.8.2") {
+    implementation("org.yaml:snakeyaml:1.33") // Needed for Airplane
+    implementation ("me.carleslc.Simple-YAML:Simple-Yaml:1.8.2") { // Needed for Airplane
         exclude(group="org.yaml", module="snakeyaml")
     }
     // Pufferfish end
 
+    implementation("com.github.technove:AIR:fe3dbb4420") // Airplane - config
+
     testImplementation("io.github.classgraph:classgraph:4.8.47") // Paper - mob goal test
     testImplementation("junit:junit:4.13.2")
     testImplementation("org.hamcrest:hamcrest-library:1.3")
