From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Fri, 18 Mar 2022 20:38:28 +1000
Subject: [PATCH] Handle projectile hits on the entity's server


diff --git a/net/minecraft/world/entity/projectile/Projectile.java b/net/minecraft/world/entity/projectile/Projectile.java
index 9f9ce6c24472b798f2ec62aba3dcb44b64be61b2..45e99efd40db98f34dec8c2a38d7fe6e0c1127e3 100644
--- a/net/minecraft/world/entity/projectile/Projectile.java
+++ b/net/minecraft/world/entity/projectile/Projectile.java
@@ -25,6 +25,9 @@ import net.minecraft.world.phys.HitResult;
 import net.minecraft.world.phys.Vec3;
 // CraftBukkit start
 import org.bukkit.projectiles.ProjectileSource;
+import puregero.multipaper.ExternalServerConnection;
+import puregero.multipaper.MultiPaperEntityInteractHandler;
+import puregero.multipaper.externalserverprotocol.ProjectileHitEntityPacket;
 // CraftBukkit end
 
 public abstract class Projectile extends Entity implements TraceableEntity {
@@ -35,6 +38,7 @@ public abstract class Projectile extends Entity implements TraceableEntity {
     public Entity cachedOwner;
     public boolean leftOwner;
     public boolean hasBeenShot;
+    private boolean hasHitEntity; // MultiPaper
 
     // CraftBukkit start
     protected boolean hitCancelled = false;
@@ -226,10 +230,22 @@ public abstract class Projectile extends Entity implements TraceableEntity {
     }
     // CraftBukkit end
 
+    public void onHit0(HitResult hitResult) { this.onHit(hitResult); }  // MultiPaper - expose onHit
     protected void onHit(HitResult hitResult) {
         HitResult.Type movingobjectposition_enummovingobjecttype = hitResult.getType();
 
         if (movingobjectposition_enummovingobjecttype == HitResult.Type.ENTITY) {
+            // MultiPaper start - handle entity hits on their server
+            if (hasHitEntity) {
+                return; // Prevent infinite hits of the entity if an external server ignores the projectile hit packet
+            }
+            hasHitEntity = true;
+            ExternalServerConnection entityOwner = MultiPaperEntityInteractHandler.getOwner(((EntityHitResult) hitResult).getEntity());
+            if (entityOwner != null) {
+                entityOwner.send(new ProjectileHitEntityPacket(this, ((EntityHitResult) hitResult).getEntity(), hitResult.getLocation()));
+                return;
+            }
+            // MultiPaper end
             this.onHitEntity((EntityHitResult) hitResult);
             this.level().gameEvent(GameEvent.PROJECTILE_LAND, hitResult.getLocation(), GameEvent.Context.of(this, (BlockState) null));
         } else if (movingobjectposition_enummovingobjecttype == HitResult.Type.BLOCK) {
