From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 20 Mar 2024 21:48:26 +0900
Subject: [PATCH] Sync entity PersistentDataContainer


diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index 9b623c451555ef97b900d4e91e90ccaa146709d0..daedce89005a13226c354ce8f5bcaec2f1f3ed6d 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -195,6 +195,7 @@ import org.bukkit.event.server.ServerLoadEvent;
 
 import co.aikar.timings.MinecraftTimings; // Paper
 import puregero.multipaper.MultiPaper;
+import puregero.multipaper.MultiPaperDirtyCraftPersistentDataContainer;
 import puregero.multipaper.MultiPaperExternalBlocksHandler;
 import puregero.multipaper.MultiPaperFileSyncer;
 import puregero.multipaper.externalserverprotocol.DifficultyUpdatePacket;
@@ -1637,6 +1638,8 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         }
         this.isIteratingOverLevels = false; // Paper
 
+        MultiPaperDirtyCraftPersistentDataContainer.tick(); // MultiPaper
+
         //this.profiler.popPush("connection"); // Purpur
         //MinecraftTimings.connectionTimer.startTiming(); // Spigot // Purpur
         this.getConnection().tick();
